eventhandling/getSliEvent.go

```diff
@@ -47,33 +47,82 @@ func (eh GetSliEventHandler) HandleEvent() error {
 
 	// send started event
 	_, err = eh.keptnHandler.SendTaskStartedEvent(eventData, utils.ServiceName)
-
 	if err != nil {
-		errMsg := fmt.Sprintf("Failed to send task started CloudEvent (%s), aborting...", err.Error())
-		log.Println(errMsg)
+		errMsg := fmt.Errorf("failed to send task started CloudEvent: %w", err)
+		log.Println(errMsg.Error())
 		return err
 	}
 
-	// create SLI Results
-	var sliResults []*keptnv2.SLIResult
+	// helper function to log an error and send an appropriate finished event
+	sendFinishedErrorEvent := func(err error) error {
+		log.Printf("sending errored finished event: %s", err.Error())
 
-	// 2: try to fetch metrics into sliResults
-	if sliResults, err = retrieveMetrics(eventData, eh.keptnHandler); err != nil {
-		// failed to fetch metrics, send a finished event with the error
-		_, err = eh.keptnHandler.SendTaskFinishedEvent(&keptnv2.EventData{
+		_, sendError := eh.keptnHandler.SendTaskFinishedEvent(&keptnv2.EventData{
 			Status:  keptnv2.StatusErrored,
 			Result:  keptnv2.ResultFailed,
 			Message: err.Error(),
 		}, utils.ServiceName)
 
-		return err
+		// TODO: Maybe log error to console
+
+		return sendError
+	}
+
+	// get prometheus API URL for the provided Project from Kubernetes Config Map
+	prometheusAPIURL, err := getPrometheusAPIURL(eventData.Project, eh.kubeClient.CoreV1())
+	if err != nil {
+		return sendFinishedErrorEvent(fmt.Errorf("unable to get prometheus api URL: %w", err))
+	}
+
+	// create a new Prometheus Handler
+	prometheusHandler := prometheus.NewPrometheusHandler(
+		prometheusAPIURL,
+		&eventData.EventData,
+		eventData.Deployment, // "canary", "primary" or "" (or "direct" or "user_managed")
+		eventData.Labels,
+		eventData.GetSLI.CustomFilters,
+	)
+
+	// get SLI queries (from SLI.yaml)
+	projectCustomQueries, err := getCustomQueries(eh.keptnHandler, eventData.Project, eventData.Stage, eventData.Service)
+	if err != nil {
+		return sendFinishedErrorEvent(
+			fmt.Errorf("unable to retrieve custom queries for project %s: %w", eventData.Project, err),
+		)
+	}
+
+	// only apply queries if they contain anything
+	if projectCustomQueries != nil {
+		prometheusHandler.CustomQueries = projectCustomQueries
+	}
+
+	// retrieve metrics from prometheus
+	sliResults := retrieveMetrics(prometheusHandler, eventData)
+
+	// If we hand any problem retrieving an SLI value, we set the result of the overall .finished event
+	// to Warning, if all fail ResultFailed is set for the event
+	finalSLIEventResult := keptnv2.ResultPass
+
+	if len(sliResults) > 0 {
+		sliResultsFailed := 0
+		for _, sliResult := range sliResults {
+			if !sliResult.Success {
+				sliResultsFailed++
+			}
+		}
+
+		if sliResultsFailed == 1 && len(sliResults) > 1 {
+			finalSLIEventResult = keptnv2.ResultWarning
```

Comment:
Not sure if this is what we want/need.
I think it should be `sliResultsFailed > 0 && sliResultsFailed < len(sliResults)`